create-calendar-collection.js

```diff
@@ -0,0 +1,280 @@
+const { existsSync } = require('node:fs');
+const fs = require('node:fs/promises');
+const path = require('path');
+
+/**
+ * The oldest event in the system is in May 2021.
+ * This pre-dates the file naming convention of the 'YYYY-MMDD' prefix.
+ * The calendar collection will begin from May 2021, and will extend until
+ * whatever is the latest month indicated by files in the Events Collection.
+ * @constant {number[]}
+ */
+const HARD_START_DATE = [2021, 5];
+
+/**
+ * Return the filenames within the specified directory treated as a list of entries of a Jekyll collection.
+ * This means that file names with an underscore prefix are omitted from the returned array.
+ * @param {string} eventsCollectionPath Path to the Events collection directory.
+ * @returns {Promise<string[]>}
+ */
+async function getEventsCollectionFileNames(eventsCollectionPath) {
+    const fileNames = await fs.readdir(eventsCollectionPath);
+
+    // Filter out filenames prefixed with an underscore as that is a Jekyll
+    // convention to omit the collection entry from the build.
+    const filteredFileNames = fileNames.filter(fileName => !fileName.startsWith('_'));
+    return filteredFileNames;
+}
+
+/**
+ * Returns a deduplicated array of the Event collection filenames mapped to an array in the format of [year, month].
+ * Any filename that does not conform to the naming convention of beginning with the event date as YYYY-MMDD will be omitted.
+ * @param {string[]} fileNames Array of Event collection filenames.
+ * @returns {number[]}
+ */
+function parseYearMonthPairsFromFileNames(fileNames) {
+    const parsedFileNames = fileNames.map(fileName => {
+        const pattern = /^(\d{4})-(\d{2})/;
+        const matches = pattern.exec(fileName);
+        if (matches) {
+            const parsedYear = Number.parseInt(matches[1], 10);
+            const parsedMonth = matches[2].startsWith('0') ?
+                Number.parseInt(matches[2].charAt(1), 10) :
+                Number.parseInt(matches[2], 10);
```

Comment:
Checking if it starts with `0` isn't adding any value. Also, this forgives bad values like `20`.

It is almost guaranteed that this piece of code will not remain as is for the next 75 years so:
```
const pattern = /^(20\d{2})-(0[1-9]|1[0-2])/;
```
This will validate the correctness of `YYYY-MM` part. However, that still doesn't do what the comment says it will because it doesn't check the `DD`. We should either change the comment to relax the criteria or apply the strict criteria by not only parsing the numbers, but also validating the date. My sugestion would be a combination:
1. relax the criteria to `YYYY-MM##@` where `#` is just some number and `@` is a non-numeric character, and
2. change the pattern to `/^(20\d{2})-(0[1-9]|1[0-2])\d{2}([^\d]|$)/`.

That said, in the past we have had complications caused by poor choice of filenames. I would much rather read the content of the file and parse the actual date from there.